File: 프로그래머스/lv1_모의고사.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> solution(vector<int> answers) {
    vector<int> answer;
    vector<int> std1 = {1,2,3,4,5};
    vector<int> std2 = {2,1,2,3,2,4,2,5};
    vector<int> std3 = {3,3,1,1,2,2,4,4,5,5};

    vector<int> score(3);
    int max = 0;

    for( int i=0; i<answers.size(); i++ ){
        /* std의 원소 갯수를 생각해서 % 연산 */
        if(answers[i] == std1[i%5]){score[0]++;}
        if(answers[i] == std2[i%8]){score[1]++;}
        if(answers[i] == std3[i%10]){score[2]++;}
    }

    if( score.empty() != true ){
        /* score 벡터에서 가장 높은 점수를 취득 */
        max = *max_element(score.begin(), score.end());
        
        /* 가장 높은 점수를 가진 사람 확인 */
        for ( int i=0; i<score.size(); i++ ){
            if( max == score[i] ){
                answer.push_back(i+1);
            }
        }
    }

    return answer;
}

int main(void){
    
    vector<int> answers = {1,3,2,4,2};
    vector<int> ret;

    ret = solution(answers);

    for( auto elem : ret ){
        cout << elem << endl;
    }

    return 0;
}

File: 프로그래머스/lv1_모의고사.py
'''
    first, second, third 의 인덱스를 나머지연산을 통해 순환 함.

    enumrate 는 인덱스와, 값을 튜플로 같이 리턴한다.
'''

def solution(answers):
    answer = []

    first = [1,2,3,4,5]
    second = [2,1,2,3,2,4,2,5]
    third = [3,3,1,1,2,2,4,4,5,5]

    score_list = []
    first_score = 0
    second_score = 0
    third_score = 0

    for idx in range (0, len(answers)):
        if( answers[idx] == first[idx%len(first)] ):
            first_score += 1
        if( answers[idx] == second[idx%len(second)] ):
            second_score += 1
        if( answers[idx] == third[idx%len(third)] ):
            third_score += 1
    
    score_list.append(first_score)
    score_list.append(second_score)
    score_list.append(third_score)

    print(score_list)
    max_num = max(score_list)

    # answer = [ idx+1 for idx, value in enumerate(score_list) if value == max_num ]
    # 아래와 같은 역할을 함
    
    for idx, value in enumerate(score_list):
        if( value == max_num):
            answer.append(idx+1)

    answer.sort()

    return answer

if __name__ == "__main__":
    answer = [1,3,2,4,2]
    ret = solution(answer)

    print(ret)